File: old-code/rarduino/src/mecanum_drive.rs
use crate::motor::Motor;

pub struct MecanumDrive {
    front_left: Motor,
    front_right: Motor,
    back_left: Motor,
    back_right: Motor,
}

impl MecanumDrive {
    pub fn new(front_left: Motor, front_right: Motor, back_left: Motor, back_right: Motor) -> Self {
        Self {
            front_left,
            front_right,
            back_left,
            back_right,
        }
    }

    // Source: Method 1 in https://www.youtube.com/watch?v=gnSW2QpkGXQ
    pub fn set_speed(&mut self, x: i16, y: i16, rotation: i16) {
        self.front_left.set_speed(y + x + rotation);
        self.front_right.set_speed(y - x - rotation);
        self.back_left.set_speed(y - x + rotation);
        self.back_right.set_speed(y + x - rotation);
    }
}

File: jarduino/drives_lib/MecanumDrive.cpp
#pragma once
#include "Motor.cpp"

// Source: Method 1 in https://www.youtube.com/watch?v=gnSW2QpkGXQ
// BY THE WAY, this works for drivetrains where the Omni wheels are on the *corners*!
class MecanumDrive {
public:
    Motor frontLeft, frontRight, backLeft, backRight;

    MecanumDrive(Motor frontLeft, Motor frontRight, Motor backLeft, Motor backRight)
        : frontLeft(frontLeft), frontRight(frontRight), backLeft(backLeft), backRight(backRight) {
    }

    void setSpeed(double x, double y, double rotation) {
        frontLeft.setSpeed(y + x + rotation);
        frontRight.setSpeed(y - x - rotation);
        backLeft.setSpeed(y - x + rotation);
        backRight.setSpeed(y + x - rotation);
    }
};